### PR TITLE
Add grad norm logging single device

### PR DESCRIPTION
**Pull request by lindawangg**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
- Addresses #1407 

#### Changelog
What are the changes made in this PR?
- Adds optional clip grad norm
- Default logs grad norm is `clip_grad_norm` is set

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

```
tune run lora_finetune_single_device --config llama2/7B_lora_single_device max_steps_per_epoch=100 clip_grad_norm=100 metric_logger._component_=torchtune.utils.metric_logging.TensorBoardLogger checkpointer.output_dir=/tmp/llama2_w_grad_norm
```
Did not observe any changes to overall compute time or loss.
* orange is with `log_grad_norm=True`, blue is baseline
image
image
image
image

```
tune run full_finetune_single_device --config llama2/7B_full_low_memory metric_logger._component_=torchtune.utils.metric_logging.TensorBoardLogger checkpointer.output_dir=/tmp/llama2_w_grad_norm_full optimizer_in_bwd=False clip_grad_norm=100
```
Observed no increase in batch/s. Verified clip_grad_norm clips gradients (in orange)
* green is baseline, purple is clip_grad_norm='inf', orange is clip_grad_norm=100
image
image
image

#### UX
- [ ] I did not change any public API;
- [x] I have added an example to docs or docstrings;

### CODE REVIEW

File: recipes/lora_finetune_single_device.py [hunk 1]
```diff
@@ -132,6 +136,10 @@ def __init__(self, cfg: DictConfig) -> None:
         self._resume_from_checkpoint = cfg.resume_from_checkpoint
         self._save_adapter_weights_only = cfg.get("save_adapter_weights_only", False)
         self._gradient_accumulation_steps = cfg.gradient_accumulation_steps
+        self._clip_grad_norm = cfg.get("clip_grad_norm", None)
+        self._log_grad_norm = cfg.get("log_grad_norm", False)
```
Comment (lindawangg) [inline recipes/lora_finetune_single_device.py:140]: We don't really need `log_grad_norm`. If `clip_grad_norm` is set, we can log grad norm by default. Computing this each step doesn't increase the time for single device.

Comment (ebsmothers) [reply recipes/lora_finetune_single_device.py:140]: Yeah based on your results I agree, seems like no negative perf impact to just logging by default. Can you also make the same changes in full_finetune_single_device.py and confirm that no negative perf impact there? (I think we should enable for both recipes in one go.)

Distributed may be a different story if we have to sync, but we can consider that out of scope for this PR.

If the results look similar for full finetune, I agree it makes sense to go ahead and remove the `log_grad_norm` config. I would even bias towards just always logging grad norm (that way you don't have to awkwardly set it to `'inf'` to turn on logging, imo that's a bit unintuitive).

**Review by ebsmothers — COMMENTED** [review]
This looks good! Apart from my comment around enabling for full finetune, we can also run a version with `clip_grad_norm` set to some finite value to ensure that we see it reflected correctly in the logs.

**Review by ebsmothers — APPROVED** [review]
This looks great, thank you!
